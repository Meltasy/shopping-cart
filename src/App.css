.productItems {
  display: flex;
  flex-wrap: wrap;
  gap: 10px;
  padding: 10px;
  justify-content: center;
}

.item {
  max-width: 300px;
  height: auto;
  padding: 10px;
  color: var(--text-color);
  background-color: var(--primary-color-light);
  border-radius: 10px;
  justify-items: center;
}

.item img {
    max-width: 280px;
    height: auto;
    object-fit: cover;
    border-radius: 10px;
}
