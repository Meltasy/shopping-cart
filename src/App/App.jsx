import { useParams } from 'react-router-dom'
import { useEffect, useState } from 'react'
import styled from 'styled-components'
import GlobalStyle from '../globalStyles'
import Navbar from '../components/Navbar/Navbar'
import Footer from '../components/Footer'
import Home from '../pages/Home/Home'
import Shop from '../pages/Shop/Shop'
import Cart from '../pages/Cart/Cart'

const AppWrapper = styled.div`
  min-height: 100vh;
  display: flex;
  flex-direction: column;
`

const MainWrapper = styled.main`
  margin-top: 3rem;
  flex: 1;
`

function App() {
  const { name } = useParams()
  const [productItems, setProductItems] = useState(null)
  const [error, setError] = useState(null)
  const [loading, setLoading] = useState(true)
  const [cartItems, setCartItems] = useState([])
  const [quantityCart, setQuantityCart] = useState(0)
  const [totalCost, setTotalCost] = useState(0)

  useEffect(() => {
    fetch('https://fakestoreapi.com/products', { mode: 'cors' })
    .then((response) => {
      if (response.status >= 400) {
        throw new Error('Server error')
      }
      return response.json()
    })
    .then((response) => {
      setProductItems(response)
    })
    .catch((error) => setError(error))
    .finally(() => setLoading(false))
  }, [])

  if (loading) return <p>Loading...</p>
  if (error) return <p>Network error</p>

  function handleAdd(itemId, price, quantity) {
    if (cartItems.find(item => item.id === itemId)) {
      setTotalCost(totalCost + price * quantity)
      setCartItems(cartItems.map(item => {
        if (item.id === itemId) {
          return { ...item, quantity: item.quantity + quantity }
        } else {
          return { ...item }
        }
      }))
    } else {
      setCartItems([ ...cartItems, { id: itemId, price: price, quantity: quantity } ])
      setQuantityCart(quantityCart + 1)
      setTotalCost(totalCost + price * quantity)
    }
  }

  function handleDelete(itemId, price, quantity) {
    setCartItems(cartItems.filter(item => item.id !== itemId))
    setQuantityCart(quantityCart - 1)
    setTotalCost(totalCost - price * quantity)
  }
  
  return (
    <>
      <AppWrapper>
        <GlobalStyle />
        <Navbar quantityCart={quantityCart} />
        <MainWrapper>
          {name === 'shop' ? (
            <Shop productItems={productItems} handleAdd={handleAdd} />
          ) : name === 'cart' ? (
            <Cart
              productItems={productItems}
              cartItems={cartItems}
              totalCost={totalCost}
              handleDelete={handleDelete}
            />
          ) : (
            <Home productItems={productItems} />
          )}
        </MainWrapper>
        <Footer />
      </AppWrapper>
    </>
  )
}

export default App
