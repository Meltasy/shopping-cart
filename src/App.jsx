import ShoppingCartProvider from './context/ShoppingCartContext'
import Home from './pages/Home'
import './App.css'

function App() {

  return (
    <ShoppingCartProvider>
      <Home />
    </ShoppingCartProvider>
  )
}

export default App
