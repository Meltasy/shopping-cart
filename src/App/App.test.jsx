import { describe, it, expect } from 'vitest'
import { render, screen } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import App from './App'


describe('App component', () => {
  it('renders App component', () => {
    const { container } = render(
      <BrowserRouter>
          <App />
      </BrowserRouter>
    )
    expect(container).toMatchSnapshot()
  })

  it('renders appropriate message while loading', async () => {
    render(
      <BrowserRouter>
          <App />
      </BrowserRouter>
    )
    expect(screen.getByText(/loading.../i)).toBeInTheDocument()
  })

  it ('renders correct heading after it has loaded', async () => {
    render(
      <BrowserRouter>
          <App />
      </BrowserRouter>
    )
    const heading = await screen.findByRole('heading', { level: 1, name: 'Galeries Libellule' })
    screen.debug()
    expect(heading).toBeInTheDocument()
  })

  it('renders all sections after it has loaded', async () => {
    render(
      <BrowserRouter>
          <App />
      </BrowserRouter>
    )
    const main = await screen.findByRole('main')
    const nav = await screen.findByRole('navigation')
    expect(main).toBeInTheDocument()
    expect(nav).toBeInTheDocument()
  })
})
