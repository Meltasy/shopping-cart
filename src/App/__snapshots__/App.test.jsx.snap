// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > renders App component 1`] = `
<div>
  <p>
    Loading...
  </p>
</div>
`;
